### PR TITLE
Hayden - Edges (C10) - API Muncher

### PR DESCRIPTION
**Pull request by halcyon-and-on**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, and how did you try querying the API? | I poked around thru their dox and also played with it in postman
What is an API Wrapper? Why is it in `lib`? How would your project change if you needed to interact with more than one API (aka more than just the Edamam API)? | an apiwrapper is kind of analogous to an activerecord model, but for accessing an external db / service. lib is where miscellaneous code files go. i did interact with more than one API and i just wrote another wrapper for it.
Describe your API wrapper, the methods you created in it, and why you decided to create those methods/how they were helpful | i have a search method and a recipe show method. i created them because thats literally what the project asked for. they helped make my project functional. the methods use CONSTANTS and some user specified args to format the proper URLs in the methods.
What was an edge case or failure case test you wrote for your API Wrapper? What was a nominal case? | I haven't finished any of the testing yet :/ I def want to test weird input like negative integers, strings for the page number, blank input, etc.
How does VCR aid in testing an API? | records an API response so you can replay the data again and again instead of wasting expensive API calls
What is the Heroku URL of your deployed application? | http://fooddd.herokuapp.com/      (be gentle with my baby!!)

### CODE REVIEW

File: test/test_helper.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+ENV["RAILS_ENV"] = "test"
+require File.expand_path("../../config/environment", __FILE__)
+require "rails/test_help"
+require "minitest/rails"require "minitest/reporters"  # for Colorized output
```
Comment (CheezItMan) [inline test/test_helper.rb:4]: You have a typo here, put a newline between `rails"` and `require`

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  <head>
+    <title>Food.</title>
+    <%= csrf_meta_tags %>
+    <%= csp_meta_tag %>
+
+    <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track': 'reload' %>
+    <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
+
+    <style>
+      body {
```
Comment (CheezItMan) [inline app/views/layouts/application.html.erb:12]: This should go in your stylesheet.

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        background-image: url(<%= @bg_image %>);
+        background-size: cover;
+      }
+    </style>
+  </head>
+
+  <body>
+    <p class="logo"><%= link_to "Food.", root_path %></p>
+    <ul class="login">
```
Comment (CheezItMan) [inline app/views/layouts/application.html.erb:21]: I would put this stuff in a `header` or other container tag.

File: lib/edamam_wrapper.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    url = BASE_URL + "app_id=#{ID}&" + "app_key=#{KEY}&" + "q=#{queery}"
+    if page
+      from = (page * 10) - 10
+      to = from + 10
+      data = HTTParty.get(url + "&from=#{from}&to=#{to}")
+    else
+      data = HTTParty.get(url)
+    end
+
+    if data.include?("Error")
```
Comment (CheezItMan) [inline lib/edamam_wrapper.rb:19]: Good that you're checking things here.

File: lib/edamam_wrapper.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    else
+      recipe = Recipe.new({"label"=> "sorry, no results for this query"})
+    end
+
+    return recipe
+  end
+
+  private
+
+  def self.recipe_squeezer(hits)
```
Comment (CheezItMan) [inline lib/edamam_wrapper.rb:56]: This method name isn't really very descriptive of what it does.

File: lib/edamam_wrapper.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    if ref.include?("http://www.edamam.com/ontologies/edamam.owl")
+      url = BASE_URL + "app_id=#{ID}&" + "app_key=#{KEY}&" + "r=#{ref}"
+    else
+      url = BASE_URL + "app_id=#{ID}&" + "app_key=#{KEY}&" + "q=#{ref}"
+    end
+
+    encoded_url = URI.encode(url)
+    data = HTTParty.get(encoded_url)
+
+    if data.include?("Error")
```
Comment (CheezItMan) [inline lib/edamam_wrapper.rb:40]: if `data.body` is [], this is going to crash.

**Comment by CheezItMan** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** |
Git hygiene |  Not very many commits and your commit messages are mostly ok, `sorry about that...` isn't very descriptive.  
Comprehension questions |  Check, No promises with your baby, you should teach her to fend for herself!  
**General** |
Rails fundamentals (RESTful routing, use of named paths) |  Somewhat, some of your routes like `/favs` and `/recipe` aren't really RESTful.   
Semantic HTML |  Pretty good
Errors are reported to the user |  Check
API Wrapper to handle the API requests |  Check
Controller testing |  **MISSING**
Lib testing | **MISSING**
Search Functionality |  Check
List Functionality |  Check
Show individual item functionality |  Check, but an invalid URI crashes the app.  
**Styling** |
List view shows 10 items at a time and/or has pagination |  Check
Attractive and usable UI |  Everything on the search results page is a **little** too big, like King Kong is a little too big. 
**API Features** |
The App attributes Edamam |**MISSING**
The VCR cassettes do not contain the API key |  **NO TESTING**
**External Resources** |
Link to deployed app on Heroku |  Check
**Overall** |  It looks like you tried to get Google OAuth working, but it's not working on Heroku.  I do like the saved recent searches.  However you should get the project requirements done **first** before going the extra mile.  I strongly suggest reviewing testing an API, both in the controller and testing the lib folder.  You hit the learning goals __except__ for testing.
